bundle-workflow/src/build_workflow/build_recorder.py

```diff
@@ -50,29 +114,79 @@ def write_manifest(self, dest_dir):
+                artifact_file, f"expected filename to include {self.component_version}"
+            )
+        with ZipFile(artifact_file, "r") as zip:
+            data = zip.read("plugin-descriptor.properties").decode("UTF-8")
+            properties = BuildRecorder.PropertiesFile(artifact_file, data)
+            properties.check_value("version", self.component_version)
+            properties.check_value("opensearch.version", self.version)
+            print(f'Checked {artifact_file} ({properties.get_value("version", "N/A")})')
+
+    def __check_maven_artifact(self, artifact_file):
+        ext = os.path.splitext(artifact_file)[1]
+        if ext not in [
+            ".jar",
+            ".asc",
+            ".md5",
+            ".sha1",
+            ".pom",
+            ".xml",
+            ".sha1",
+            ".sha256",
+            ".sha512",
+            ".module",
+            ".zip",
+            ".war",
+        ]:
+            raise BuildRecorder.ArtifactInvalidError(
+                artifact_file, f"{ext} is not a valid extension for a maven file"
+            )
+        if os.path.splitext(artifact_file)[1] == ".jar":
+            with ZipFile(artifact_file, "r") as zip:
+                data = zip.read("META-INF/MANIFEST.MF").decode("UTF-8")
+                properties = BuildRecorder.PropertiesFile(artifact_file, data)
+                properties.check_value_in(
+                    "Implementation-Version",
+                    [self.component_version, self.opensearch_version, None],
+                )
+                print(
+                    f'Checked {artifact_file} ({properties.get_value("Implementation-Version", "N/A")})'
+                )
+
     class BuildManifestBuilder:
         def __init__(self, build_id, name, version, arch, snapshot):
             self.data = {}
             self.data["build"] = {}
             self.data["build"]["id"] = build_id
             self.data["build"]["name"] = name
-            self.data["build"]["version"] = (
-                str(version) + "-SNAPSHOT" if snapshot else str(version)
-            )
+            self.data["build"]["version"] = version
             self.data["build"]["architecture"] = arch
             self.data["build"]["snapshot"] = str(snapshot).lower()
             self.data["schema-version"] = "1.0"
             # We need to store components as a hash so that we can append artifacts by component name
             # When we convert to a BuildManifest this will get converted back into a list
             self.data["components_hash"] = {}
 
-        def append_component(self, name, repository_url, ref, commit_id):
+        def append_component(self, name, version, repository_url, ref, commit_id):
             component = {
                 "name": name,
                 "repository": repository_url,
                 "ref": ref,
                 "commit_id": commit_id,
                 "artifacts": {},
+                "version": version,
```

Comment:
along similar lines as previous comment, does it make sense to have a `qualifier` (e.g. snapshot, rc, etc.) defined separately?

Reply:
This is the actual component version as written into the JAR. We can separate a qualifier as part of https://github.com/opensearch-project/opensearch-build/issues/385